feat: Implement sequence processing improvements for SHACL Node Expressions (Issue #484)

simonstey opened the pull request:
This pull request implements comprehensive changes to address [Issue #484](https://github.com/w3c/data-shapes/issues/484) regarding sequence processing naming inconsistencies in SHACL Node Expressions. The changes align the vocabulary and documentation with the sequence-based nature of node expression processing while maintaining backward compatibility through deprecation notices.

## Problem Statement

SHACL Node Expressions are fundamentally sequence-based but were using set-style operation names, creating confusion:
- Operations like `union` and `minus` suggested set semantics despite working on ordered sequences
- The generic `path` property conflicted conceptually with constraint `sh:path` 
- Missing advanced sequence operations limited processing capabilities

## Solution

### 1. Vocabulary Renaming for Sequence Semantics

**Renamed Operations:**
- `shnex:union` → `shnex:join` - Emphasizes sequence concatenation with order preservation
- `shnex:minus` → `shnex:remove` - Clearer operation name for sequence subtraction
- `shnex:path` → `shnex:pathValues` - Distinguishes from constraint `sh:path` (needs to align with #514 )

**Deprecation Strategy:**
- Updated `shacl.ttl` with deprecation notices for `sh:union` and `sh:minus`
- Maintained backward compatibility while guiding migration to new terms

### 2. Advanced Sequence Operations

**New Operations Added:**
- `shnex:flatMap` - Applies expression to each input node and flattens results
- `shnex:findFirst` - Returns first node conforming to a given shape
- `shnex:matchAll` - Returns true if all nodes conform to a given shape

## Files Modified

### Vocabulary Files
- **`shacl12-vocabularies/shnex.ttl`**
  - Added complete RDF definitions for FlatMap, FindFirst, and MatchAll expressions
  - Updated existing definitions with sequence-appropriate naming
  - Enhanced property comments for clarity

- **`shacl12-vocabularies/shacl.ttl`**
  - Added deprecation notices for `sh:union` and `sh:minus`
  - Clear migration guidance to new sequence-based terms

### Documentation
- **`shacl12-node-expr/index.html`**
  - Renamed sections: UnionExpression → JoinExpression, MinusExpression → RemoveExpression


### New Advanced Operations:
```turtle
# Find first senior employee
sh:values [
    shnex:findFirst [
        shnex:nodes [ shnex:pathValues ex:employee ] ;
        shnex:findFirst ex:SeniorEmployeeShape ;
    ] ;
] .

# Check if all employees are active
sh:values [
    shnex:matchAll [
        shnex:nodes [ shnex:pathValues ex:employee ] ;
        shnex:matchAll ex:ActiveEmployeeShape ;
    ] ;
] .
```
---

Closes #484 

* [See this document rendered online here](https://raw.githack.com/w3c/data-shapes/issue-484-sequence-processing/shacl12-node-expr/index.html)

nicholascar committed:
what's new, cribbed from LinkedIn and extended

nicholascar committed:
common specifications listing

nicholascar committed:
local biblio test

nicholascar committed:
add common spec list to all docs; add light/dark mode option

ajnelson-nist committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

ajnelson-nist committed:
Update shacl12-overview/index.html

Committing for grammar remarks, but this does not visit the topic raised about whether there officially exists a "SHACL 1.1".

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

ajnelson-nist committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

ajnelson-nist committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

ajnelson-nist committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

ajnelson-nist committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

ajnelson-nist committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

simonstey committed:
feat(shnex): introduce advanced sequence manipulation expressions (flatMap, findFirst, matchAll)

simonstey committed:
refactor(shnex): rename shnex:union and shnex:minus to shnex:join and shnex:remove in spec, examples, and vocabulary

simonstey committed:
refactor(shnex): rename shnex:path to shnex:pathValues throughout specification and examples

simonstey committed:
refactor(shnex): rename shnex:union and shnex:minus to shnex:join and shnex:remove in spec, examples, and vocabulary

simonstey committed:
feat: :sparkles: Merge latest changes from upstream branch; add sequence based expressions

nicholascar committed:
Merge pull request #520 from w3c/whats-new

SHACL Overview, what's new

nicholascar committed:
Merge branch 'gh-pages' into common-spec-elements

nicholascar committed:
Update shacl12-common/local-biblio.js

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

HolgerKnublauch committed:
#505: Added section on Dynamic SHACL, including two examples (#509)

* #505: Added section on Dynamic SHACL, including two examples

* Apply suggestions from code review

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

* Apply suggestions from code review

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

* Update shacl12-node-expr/index.html

Co-authored-by: David Habgood <dcchabgood@gmail.com>

* Update shacl12-node-expr/index.html

Co-authored-by: David Habgood <dcchabgood@gmail.com>

* Update shacl12-node-expr/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

---------

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>
Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>
Co-authored-by: David Habgood <dcchabgood@gmail.com>

TallTed commented on shacl12-node-expr/index.html:
```diff
@@ -1117,6 +1117,15 @@ <h3>Path Expressions</h3>
+					<div class="note">
+						<p><strong>Important:</strong> Note the distinction between <code>sh:path</code> and <code>shnex:pathValues</code>:</p>
+						<ul>
+							<li><code>sh:path</code> is used in <strong>property shapes</strong> to specify which property path should be constrained during validation. It defines what property or path the shape's constraints apply to.</li>
```
```suggestion
							<li><code>sh:path</code> is used in <strong>property shapes</strong> to specify the property path that will be constrained during validation. It defines the property or path to which the shape's constraints apply.</li>
```

TallTed commented on shacl12-node-expr/index.html:
```diff
@@ -1117,6 +1117,15 @@ <h3>Path Expressions</h3>
+						<p><strong>Important:</strong> Note the distinction between <code>sh:path</code> and <code>shnex:pathValues</code>:</p>
+						<ul>
+							<li><code>sh:path</code> is used in <strong>property shapes</strong> to specify which property path should be constrained during validation. It defines what property or path the shape's constraints apply to.</li>
+							<li><code>shnex:pathValues</code> is used in <strong>node expressions</strong> to specify which property path should be traversed to generate a sequence of values. It produces the actual values found by following the path.</li>
```
```suggestion
							<li><code>shnex:pathValues</code> is used in <strong>node expressions</strong> to specify the property path that will be traversed to generate a sequence of values. It produces the actual values found by following the path.</li>
```

TallTed commented on shacl12-node-expr/index.html:
```diff
@@ -1521,21 +1530,21 @@ <h3>Minus Expressions</h3>
-							The <a>output nodes</a> of the <a>minus expression</a> are the <a>nodes</a> in <code>N</code>
+							The <a>output nodes</a> of the <a>remove expression</a> are the <a>nodes</a> in <code>N</code>
 							except those that are also in <code>M</code>, preserving the order of <code>N</code>.
 							Nodes must be equal using <a>term equality</a>, i.e. <code>"01"^^xsd:integer</code> is distinct from <code>"1"^^xsd:integer</code>.
```
```suggestion
							Nodes must be equal using <a>term equality</a>, i.e., <code>"01"^^xsd:integer</code> is distinct from <code>"1"^^xsd:integer</code>.
```

TallTed reviewed (changes requested):
Small stuff

HolgerKnublauch committed:
#514: Renamed shnex:path to shnex:pathValues and switched from shnex:nodes to shnex:focusNode (#525)

* #514: Renamed shnex:path to shnex:pathValues and shnex:nodes to shnex:focusNode, added example

* Finished renaming, clean up

nicholascar committed:
Merge pull request #521 from w3c/common-spec-elements

Common spec elements

nicholascar committed:
improve the way original SHACL is mentioned

nicholascar committed:
Profiles of and Profiling with starts

ajnelson-nist committed:
Fix typo

nicholascar committed:
Update shacl12-sparql/index.html

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

nicholascar committed:
Update shacl12-core/index.html

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

nicholascar committed:
Update shacl12-overview/index.html

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

edmondchuc committed:
add ui doc skeleton

HolgerKnublauch committed:
#551: Clarified evalExpr parameters (#552)

edmondchuc committed:
Update shacl12-ui/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

edmondchuc committed:
Update shacl12-ui/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

edmondchuc committed:
Update shacl12-ui/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

edmondchuc committed:
Update shacl12-ui/index.html

Co-authored-by: Thomas Bergwinkl <bergos@users.noreply.github.com>

edmondchuc committed:
update company url

bergos committed:
added sh:conformanceDisallows, updated conformance-definition

nicholascar committed:
2015 -> 2017, SHACL 2015 -> SHACL Original

nicholascar committed:
Merge pull request #548 from w3c/no-shacl-1.0

Make references to original SHACL consistent

edmondchuc committed:
Update shacl12-ui/index.html

Co-authored-by: Ieben Smessaert <smessie@users.noreply.github.com>

edmondchuc committed:
Update shacl12-ui/index.html

Co-authored-by: Ieben Smessaert <smessie@users.noreply.github.com>

HolgerKnublauch committed:
#178: Added sh:someValue (#524)

Also closing #485 while I was there

bergos committed:
improved conformance definition language, fixed tabs vs. spaces

bergos committed:
added not about def conformance disallow set, updated severity section

edmondchuc committed:
Merge pull request #550 from w3c/shacl-ui-init

add ui doc skeleton

bergos committed:
Apply suggestions from code review

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

bergos committed:
improved naming for conformance-disallow set, describe default set

edmondchuc committed:
Update readme to draft profile and ui docs

nicholascar committed:
Merge pull request #561 from w3c/update-links

Update readme to draft profile and ui docs

HolgerKnublauch committed:
Issue 490 images updated (#531)

* #490: Updated images, also clarified one TODO in the Getting Started section

* Fixed a broken reference

HolgerKnublauch committed:
#501 (#557)

* #501

* Added simple examples to illustrate the three cases

* Update shacl12-node-expr/index.html

Co-authored-by: Andy Seaborne <andy@apache.org>

* Fixed grammar

---------

Co-authored-by: Andy Seaborne <andy@apache.org>

HolgerKnublauch committed:
#500 (#556)

* #500

* Update shacl12-node-expr/index.html

Co-authored-by: David Habgood <dcchabgood@gmail.com>

* Applied Andy's feedback

---------

Co-authored-by: David Habgood <dcchabgood@gmail.com>

nicholascar committed:
improve prov-constraints note

nicholascar committed:
Merge remote-tracking branch 'origin/pro-of-sh-sects' into pro-of-sh-sects

HolgerKnublauch committed:
#498: Started with examples (#513)

* #498: Started with examples

* Progress

* Generalized scope to be node -> node map, added shnex:arg to use scope

* Progress

* Finished first pass

* Clarified that list functions return at most one node

* Fixed shnex:path -> pathValues renaming in examples

* Added links to examples

nicholascar committed:
fix ref

nicholascar committed:
first pass Core Priv & Sec section + extension for Profiling

nicholascar committed:
first pass roles subsection

smessie committed:
chore(ui): Add myself (Ieben) as editor of SHACL UI

nicholascar committed:
Merge pull request #564 from w3c/smessie-shui-editors

Add myself (Ieben) as editor of SHACL UI

afs committed:
GH-567: Rename SHACL Inference Rules as SHACL Rules

bergos committed:
Update shacl12-core/index.html

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

bergos committed:
added conformanceDisallows to changes section

bergos committed:
fixed whitespace

bergos committed:
copied datashapes.org widgets to ui spec

bergos committed:
fixed widget anchor links

ajnelson-nist committed:
Add tests demonstrating conformanceDisallows

Test contents adapted from Core document example by @bergos.

Test structure adapted from this file, picked more or less at random:
shacl12-test-suite/tests/core/path/path-alternative-001.ttl

Signed-off-by: Alex Nelson <alexander.nelson@nist.gov>

ajnelson-nist committed:
#570: Resolve test note (#571)

Signed-off-by: Alex Nelson <alexander.nelson@nist.gov>

bergos committed:
updated dash namespaces to shui

bergos committed:
removed TQ specific text, small cleanup

HolgerKnublauch committed:
Issue 559: Added sh:codeIdentifer annotation property for code generators (#566)

* #559: Added sh:fieldName

* Generalized to all shapes and renamed to sh:codeIdentifier

* Removed <a> around the term local name

* Fixed grammar (unrelated to this specific PR)

* Apply suggestions from code review

Co-authored-by: Andy Seaborne <andy@apache.org>

* Updated header to include node shapes

---------

Co-authored-by: Andy Seaborne <andy@apache.org>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

YoucTagh committed:
#444 Add Getting Started Section (#507)

* add getting started section

* Suggested edits

- Removed use of abbreviations such as "isn't" - written English usually should prefer the longer form "is not"
- Extended example to include class definition to clarify SHACL isn't limited to shapes
- Deleted the old example section but moved the terminology introduction down
- Avoid the term "rule" which is overloaded. Instead always use "constraint" to not clash with the use of rule in inference rules
- Use property shapes with URI and type instead of blank nodes, yet added a section to explain the syntactic variations

* Apply suggestions from code review

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

* Update shacl12-core/index.html

Co-authored-by: Thomas Bergwinkl <bergos@users.noreply.github.com>

* Update shacl12-core/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

* Update shacl12-core/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

---------

Co-authored-by: Holger Knublauch <holger@topquadrant.com>
Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>
Co-authored-by: Thomas Bergwinkl <bergos@users.noreply.github.com>

smessie committed:
Moved CSS styling to `shacl12-common/style.css` and improved dark mode support (#577)

* Move CSS styling to `shacl12-common/style.css`

* Improved dark mode support

smessie committed:
Improved dark mode support for UML and `.syntax` (#579)

nicholascar committed:
apply new short IRIs

HolgerKnublauch committed:
Fixed broken references to <dfn> definitions (#578)

HolgerKnublauch committed:
#580: Deleted JSON-LD and SHACLC from core and node expr (#583)

Also cleaned up indentation of <tbody> blocks

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
Update shacl12-profiling/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
I think the use of "base specification" is redundant, so I've removed it all togther

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

nicholascar committed:
update as per @ajnelson-nist suggestion in review comment

nicholascar committed:
Merge pull request #549 from w3c/pro-of-sh-sects

Profiles of and Profiling with starts

HolgerKnublauch committed:
Issue 506 (#585)

* #506: Started replacing blank node property shapes

* #506: More named property shapes

* Apply suggestions from code review

Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>

---------

Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>

YoucTagh committed:
Update shacl12-core/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

YoucTagh committed:
Update shacl12-core/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

YoucTagh committed:
Update shacl12-core/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

HolgerKnublauch committed:
Issue 517 (#574)

* #517: Added sh:shape property (as target)

* Added declaration of sh:shape property

* Added sh:targetWhere

* Added a performance warning

* Update shacl12-core/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

* Added test cases

* Also mention sh:shape in target definition

* Update shacl12-test-suite/tests/core/targets/targetWhere-001.ttl

Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>

* Update shacl12-core/index.html

Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>

* Update shacl12-core/index.html

Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>

* Update shacl12-core/index.html

Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>

* Update shacl12-core/index.html

Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>

* Update shacl12-core/index.html

Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>

* Update shacl12-vocabularies/shacl.ttl

Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

---------

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>
Co-authored-by: Yousouf Taghzouti <48333491+YoucTagh@users.noreply.github.com>
Co-authored-by: Alex Nelson <alexander.nelson@nist.gov>

HolgerKnublauch committed:
#588: Fixed order of definitions so that shacl-subclass and shacl-type is used first (#589)

bergos committed:
Apply suggestions from code review

Co-authored-by: Ieben Smessaert <smessie@users.noreply.github.com>
Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>
Co-authored-by: Edmond Chuc <37032744+edmondchuc@users.noreply.github.com>

nicholascar committed:
Merge remote-tracking branch 'origin/gh-pages' into gh-pages

siwerner committed:
Update index.html

I updated my (Simon Werner) company name, address, and email.

nicholascar committed:
Merge pull request #598 from w3c/siwerner-patch-1

Update index.html

nicholascar committed:
Merge remote-tracking branch 'origin/gh-pages' into gh-pages

HolgerKnublauch commented:
@simonstey this will be a great contribution, but the PR has been in draft state and unchanged for several weeks. Do you need input/help to finish it?

bergos committed:
change ui widget score to ul lists

nicholascar committed:
Merge pull request #601 from w3c/short-iris

Short iris

bergos committed:
improved wording

bergos committed:
Update shacl12-ui/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

HolgerKnublauch committed:
#606: Avoid duplicate data-validator (#607)

HolgerKnublauch committed:
#592: Fixed typo (#593)

smessie committed:
Only show language in tab group if its div exists (#616)

bergos committed:
Update shacl12-ui/index.html

Co-authored-by: Ieben Smessaert <smessie@users.noreply.github.com>

bergos committed:
Update shacl12-ui/index.html

Co-authored-by: Ieben Smessaert <smessie@users.noreply.github.com>

bergos committed:
Update shacl12-ui/index.html

Co-authored-by: Ieben Smessaert <smessie@users.noreply.github.com>

bergos committed:
Update shacl12-ui/index.html

Co-authored-by: Ieben Smessaert <smessie@users.noreply.github.com>

bergos committed:
made letter after rendering capital in ui spec

HolgerKnublauch committed:
#614: Attempt to work around outdated W3C validator complaints (#621)

nicholascar committed:
move to local copies of CSS files (#628)

HolgerKnublauch committed:
Added style.css to core too

nicholascar committed:
Merge pull request #631 from w3c/issue-626-core

Added style.css to core too

smessie committed:
Issue 534: Support for property paths in SHACL UI (#600)

* Define support for property paths in SHACL UI

* Remove targets from the examples

Apply suggestions from code review

Co-authored-by: Thomas Bergwinkl <bergos@users.noreply.github.com>

* Add notes on restrictions and challenges

* Add shape example to "Other Complex Paths" subsection

* Don't put RFC2119 keywords in bold

* Replace repetitive RFC2119 keywords

---------

Co-authored-by: Thomas Bergwinkl <bergos@users.noreply.github.com>

HolgerKnublauch committed:
#629 (#630)

bergos committed:
Update shacl12-ui/index.html

Co-authored-by: Ieben Smessaert <smessie@users.noreply.github.com>

bergos committed:
Merge pull request #572 from w3c/issue-537-ui-widgets

Issue 537 UI widgets

HolgerKnublauch committed:
#209: Added clarifying paragraph that node shapes should use rdfs:label/comment (#632)

bergos committed:
Merge branch 'gh-pages' into issue-453

bergos committed:
Merge pull request #555 from w3c/issue-453

added sh:conformanceDisallows, updated conformance-definition

simonstey committed:
Apply suggestion from @TallTed

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

simonstey committed:
Apply suggestion from @TallTed

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

simonstey committed:
Apply suggestion from @TallTed

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

simonstey committed:
Apply suggestion from @TallTed

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

simonstey committed:
Apply suggestion from @TallTed

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

simonstey committed:
Apply suggestion from @TallTed

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

simonstey committed:
Add advanced sequence operations section with FlatMap, FindFirst, and MatchAll expressions

simonstey committed:
Clarify path expression terminology and add important distinction note between sh:path and shnex:pathValues

simonstey committed:
refactor(shnex): rename Minus and Union expressions to Remove and Join in vocabulary

simonstey commented:
@TallTed @HolgerKnublauch I now rebased this PR onto gh-pages (and hopefully haven't broken anything in the process of doing so)

@TallTed I've applied your 3 suggestions, thx!

HolgerKnublauch commented:
Thanks @simonstey! The branch seems to be on the wrong base though - it displays 115 changed files and 147 commits. Maybe the easiest way forward is to create a new branch that only overwrites the node-expr doc?

simonstey reviewed (commented):
yep, I guess that's the better approach

simonstey commented:
I'll close this PR in favor of https://github.com/w3c/data-shapes/pull/636
